posts/simp-made-simple.md

```diff
@@ -0,0 +1,343 @@
+
+3) The second monad transformer application: `ReaderT Simp.Context $ StateRefT Simp.State MetaM`. 
+  The `SimpM` monad should also be able to access the "context" that `simp` is running in, e.g. which simp theorems it has access to and so on. This is captured by the type `Simp.Context`.
+  Here, the situation is not quite the same as when we were adding a `Simp.State` state to `MetaM`: while we will often want to change the state during the `simp` call, we don't expect the context to need to change. In programmer lingo, the context should be _immutable_.
+  Thus, we use a different monad transformer called `ReaderT`, which is almost identical to `StateT`, but outputs a new monad where one can only read the type passed as parameter. 
+    > For completness: when working with `ReaderT`, one can still locally overide the
+    > value of the variable that the monad keeps track of by using `withReader`. Intuitively, 
+    > the difference between `State(Ref)T` and `ReaderT` is the following: 
+    > - In `State(Ref)T`, one has access to a global variable that can be modified at 
+    >   will, 
+    > - In `ReaderT`, given a program `x : ReaderT a m`, one can only chose to *execute* `x` with
+    >   a different context. In particular, the context before and after the execution of `x` stays the same.
+
+4) The final monad transformer application: `ReaderT Simp.MethodsRef $ ReaderT Simp.Context $ StateRefT Simp.State MetaM`. 
+  This outputs a monad that has access to `Simp.Method` (passed via a ref). 
+  This capture the "pre" and "post" procedures that `simp` can use, as well as the discharger that `simp` can use, etc.
+
+## `Simproc`s
+
+We can finally define what a `simproc` is formally. 
+Recall that intuitively, a simproc takes in an expression and outputs a simplification step, possibly after modifying the current `SimpM` state (e.g. by adding new goals to be closed by the discharger).
+This behavior is formally encapsulated by the [`Simproc`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=Lean.Meta.Simp.Simproc#doc) type:
+```lean
+abbrev Simproc  := Expr → SimpM  Step
+```
+
+Concretely, it is helpful to think of a simproc as a function (or rather metaprogram) of the form
+```lean
+def mySimproc (e : Expr) : SimpM Step := do
+  -- Various manipulations involving the expression `e`
+  ...
+  let step : Step := ...
+  ...
+  return step
+```
+
+The picture above is however a slight oversimplification, as `simp` does not consume bare elements of type `Simproc`. 
+Instead, a simproc is an element of type `Simproc` annotated with the extra data mentioned in the overview subsection, like whether the simproc is `pre` or `post`, and what kind of expression it matches on. 
```

Comment:
I must say it's confusing that a simproc isn't just an element of type `Simproc`. I expect people will read the whole paragraph before realising. Could we have a warning much earlier, or something like that?